tutorial/lesson_24_async.cpp

```diff
@@ -0,0 +1,241 @@
+        //             // Acquire a semaphore or block until the data from producer is ready.
+        //             // The semaphore is released by producer thread, when the data was fully
+        //             // produced.
+        //             acquire(producer1.semaphore, 1)
+        //             consume producer1 {
+        //                 consumer[_, _, c] = <computations which use producer[_, _, c % 2]>
+        //                 // Release a semaphore to indicate that plane was consumed, producer will
+        //                 // acquire this semaphore in the other thread.
+        //                 release(producer1.folding_semaphore)
+        //             }
+        //         }
+        //     }
+        // }
+        consumer.realize({128, 128, 4});
+    }
+
+    {
+        // In the previous example, we relied on storage folding to to use double buffering
+        // technique, but there is another, more direct way to express that.
+        Func producer("producer"), consumer("consumer");
+
+        producer(x, y, c) = (x + y) * (c + 1);
+        consumer(x, y, c) = producer(x - 1, y - 1, c) + producer(x, y, c) + producer(x + 1, y + 1, c);
+
+        consumer.compute_root();
+
+        // As mentioned in the previous example, the planes are processed separately, so producer can run
+        // ahead and start producing plane `c + 1`, while `consumer` consumes already produced plane `c`.
+        // A more direct way to express this would be to hoist storage of `producer` to ouside of the loop
+        // `c` over planes, double its size and add necessary indices to flip the  planes.
+        // The first part can be achieved with `hoist_storage` directive and the rest is done with 
+        // `ring_buffer`. This schedule is rougly equivalent to the schedule in the previous example.
+        producer
+            .async()
+            .compute_at(consumer, c)
+            .hoist_storage(consumer, Var::outermost())
+            .ring_buffer(2);
+
+        // The high-level structure of the generated code will be very similar to the previous example.
+        consumer.realize({128, 128, 4});
+    }
+
+    {
+        // The advantage of the `hoist_storage` + `ring_buffer` approach is that it can be applied to
+        // fairly arbitrary loop splits and tilings. For example, in the following schedule instead of 
+        // double buffering over whole planes, we double buffer over sub-regions or tiles of the planes.
+        // This is not possible to achieve with fold_storage, because it works over the *storage*
+        // dimensions of the function and not the loop splits.
+        Func producer("producer"), consumer("consumer");
+
+        producer(x, y, c) = (x + y) * (c + 1);
+        consumer(x, y, c) = producer(x - 1, y - 1, c) + producer(x, y, c) + producer(x + 1, y + 1, c);
+
+        consumer.compute_root()
+            .tile(x, y, xo, yo, xi, yi, 16, 16, TailStrategy::Auto);
+
+        producer
```

Comment:
Suggestion: It would be useful to show what the storage and execution-time of each schedule is.

Reply:
I've added a couple of notes that storage is going to be doubled across the tutorial.